core/shared/src/main/scala-dotty/zio/NeedsEnv.scala

```diff
@@ -16,13 +16,22 @@
 
 package zio
 
-trait NotExtends[A, B] extends Serializable
+import scala.annotation.implicitNotFound
+import scala.implicits.Not
 
-object NotExtends {
-  implicit def notExtends0[A, B]: A NotExtends B      = new NotExtends[A, B] {}
-  implicit def notExtends1[A <: B, B]: A NotExtends B = ???
-  @annotation.implicitAmbiguous(
-    "The environment ${A} already contains service ${B}, are you sure you want to overwrite it? Use Has#update to update a service already inside the environment."
-  )
-  implicit def notExtends2[A <: B, B]: A NotExtends B = ???
+/**
+ * A value of type `NeedsEnv[R]` provides implicit evidence that an effect with
+ * environment type `R` needs an environment, that is, that `R` is not equal to
+ * `Any`.
+ */
+@implicitNotFound(
+  "This operation assumes that your effect requires an environment. " +
+    "However, your effect has Any for the environment type, which means it " +
+    "has no requirement, so there is no need to provide the environment."
+)
+sealed trait NeedsEnv[+R]
```

Comment:
This trait + object + implicit def combo looks pretty involved, instead I think you could do:
```scala
type NeedsEnv[+R] = Not[R =:= Any]
```
(this can be at the top-level since dotty supports top-level definitions, or wrapped in a package object)

Reply:
@smarter If we do that is there a way to provide a custom error message when the implicit is not found like we do today? I think that is quite important.

Reply:
`@implicitNotFound` works on a type alias: https://scastie.scala-lang.org/DSpKypbqRT6mXAjM7ly8YA (hmm, looks like we should disable the automatic import suggestions when implicitNotFound is used).

Reply:
Okay great!

Reply:
So playing with this more I don't know if it is going to work because of variance. We need the variance on `CanFail` and `NeedsEnv` so they work with all of our data types and methods that use variance extensively. But if it is just a type alias we need to have the same variance as `Not` which is invariant.

Reply:
Can you give an example where you need it to be variant?